genic/thermal.c

```diff
@@ -76,29 +74,29 @@ init_thermalvel(const double v_amp, double max_fd,const double min_fd)
 
     gsl_integration_workspace_free (w);
 
-    double total_frac = fermi_dirac_cumprob[LENGTH_FERMI_DIRAC_TABLE-1]/total_fd;
+    double total_frac = thermals->fermi_dirac_cumprob[LENGTH_FERMI_DIRAC_TABLE-1]/total_fd;
     //Normalise total integral to unity
     for(i = 0; i < LENGTH_FERMI_DIRAC_TABLE; i++)
-        fermi_dirac_cumprob[i] /= fermi_dirac_cumprob[LENGTH_FERMI_DIRAC_TABLE - 1];
+        thermals->fermi_dirac_cumprob[i] /= thermals->fermi_dirac_cumprob[LENGTH_FERMI_DIRAC_TABLE - 1];
 
     /*Initialise the GSL table*/
-    fd_intp = gsl_interp_alloc(gsl_interp_cspline,LENGTH_FERMI_DIRAC_TABLE);
-    fd_intp_acc = gsl_interp_accel_alloc();
-    gsl_interp_init(fd_intp,fermi_dirac_cumprob, fermi_dirac_vel,LENGTH_FERMI_DIRAC_TABLE);
+    thermals->fd_intp = gsl_interp_alloc(gsl_interp_cspline,LENGTH_FERMI_DIRAC_TABLE);
+    thermals->fd_intp_acc = gsl_interp_accel_alloc();
+    gsl_interp_init(thermals->fd_intp,thermals->fermi_dirac_cumprob, thermals->fermi_dirac_vel,LENGTH_FERMI_DIRAC_TABLE);
     return total_frac;
 }
 
 /*Add a randomly generated thermal speed in v_amp*(min_fd, max_fd) to a 3-velocity*/
 void
-add_thermal_speeds(float Vel[])
+add_thermal_speeds(struct thermalvel * thermals, float Vel[])
 {
-    const double p = gsl_rng_uniform (g_rng);
+    const double p = gsl_rng_uniform (thermals->g_rng);
```

Comment:
Where did you set the seed?

Reply:
I didn't - I just used the default seed. Shall I set it?

Reply:
It will be correlated on different ranks, which is a very bad thing. 
A quick fix is to seed it by the rank -- but you probably want something that doesn't depend on the number of ranks. Main Gadget use a table of random numbers for different particle ID (by the modulus). The draw back is that it will have some spatial correlation. We probably need better scheme, e.g. generate a 2d seed table, then in each pencil sample the random number from the 2d. Some refactoring of the current gaussian white noise code would do. But too hairy?

Reply:
A good point. Also affect neutrino particle simulations with MP-GenIC. I will fix it.